### DISCUSSION TITLE
Behaviour of ^C (CTRL-C) handling

### DISCUSSION THREAD

**Discussion by prologic** [mvdan/sh#915] [Q&A] [answered]
Hey 👋

Just want to clarify how `^C` (CTRL-C) is handled here by this library? I'm using it as a shell (of sorts) in [GoNix](https://git.mills.io/prologic/gonix) as a busybox-style applet.

Even outside of GoNix itself I observe the following behaviour:

```console
(⎈ |local:default)
prologic@Jamess-iMac
Thu Sep 15 13:57:08
~/tmp/sh
 (master) 0
$ ./gosh
$ ^C
(⎈ |local:default)
prologic@Jamess-iMac
Thu Sep 15 13:57:11
~/tmp/sh
 (master) 130
$ echo $?
130
```

- Hitting `^C` here terminates the shell with an exit status of `130`

In say a Bash shell, `^C` does not terminate the current shell (as there is no "program" in the foreground) but `$?` does have the value of `130` hmmm?

Some context, to this question. I'm actually trying to debug some missing or bad signal handling where-by hitting `^C` incorrectly terminates my shell or ssh session and not the program I happen to have running in the foreground of my shell (using this library).

Thoughts?

**Comment by mvdan** [comment, marked as answer]
I assume this is a question for cmd/gosh, the main program that implements a fairly basic interactive shell?

There's a fairly easy answer here: we don't set up any signal handlers, so ^C has its default behavior, which is to kill the entire process :) We should probably teach gosh to do the right thing instead, which seems to be to discard the in-progress line and start a new line.

**Reply by prologic** [reply]
Agreed 👌

**Reply by mvdan** [reply]
Happy to review a PR :) I think we can do without a test for now, because it's slightly tricky to test (we would need a separate process), and we currently only test runInteractive. But of course also happy to review an end-to-end test.

If you don't have time for a PR then that's also fine - then we should move this to the issue tracker.

**Comment by prologic** [comment]
@mvdan Would you accept a patch like the one here:

https://git.mills.io/prologic/gonix/commit/d8755bba8d09259551280134399faa40ad8e7e9a

if so I'll submit a PR as I've basically just borrowed most of your `cmd/gosh` code anyway since I could not really "import" it into my project as an Applet (and you probably don't want that anyway 😅)

**Reply by mvdan** [reply]
Yep, something like that looks right! And yup, cmd/gosh is of course open source, but it's not really meant to be a fully featured shell. If someone wanted to contribute such a fancy shell (with readline, autocompletion, etc) that could always be an option, though it would likely have to be a separate Go module to not add unnecessary dependencies to the main module.

**Reply by prologic** [reply]
Yeah right now that's pretty much all I'm missing from a traditional POSIX(ish) shell:

- Auto Complete
- Readline support
- History

What else do you want? 🤔

**Reply by prologic** [reply]
I don't mind spinning this into a separate project, but it will have a `Main(w io.Writer, args []string) error` to conform with GoNix'x Applet interface(s)

**Reply by mvdan** [reply]
That list sounds right. I think the main tricky part is the API, because some users may want a significant amount of control over certain features like autocompletion or history.

**Reply by prologic** [reply]
FWIW I _tried_ to integrate the [elvish](https://github.com/elves/elvish) shell into my project, but it turns out it wants to have a database and a daemon just to store history 😳 It also turns out that this is rather controversial 🤔 So I've switched back to your project, added the appropriate signal handling and I'm quite happy so far 👌

As the need arises I _might_ spawn a new `sh` project with the above mentioned features based on imports of your library 👌
